utils/src/main/java/software/amazon/awssdk/utils/async/InputStreamSubscriber.java

```diff
@@ -120,12 +124,14 @@ public int read(byte[] bytes, int off, int len) {
 
     @Override
     public void close() {
-        if (inputStreamState.compareAndSet(State.UNINITIALIZED, State.CLOSED)) {
-            delegate.onSubscribe(new NoOpSubscription());
-            delegate.onError(new CancellationException());
-        } else if (inputStreamState.compareAndSet(State.READABLE, State.CLOSED)) {
-            subscription.cancel();
-            onError(new CancellationException());
+        synchronized (subscribeLock) {
+            if (inputStreamState.compareAndSet(State.UNINITIALIZED, State.CLOSED)) {
+                delegate.onSubscribe(new NoOpSubscription());
+                delegate.onError(new CancellationException());
+            } else if (inputStreamState.compareAndSet(State.READABLE, State.CLOSED)) {
+                subscription.cancel();
```

Comment:
If `close()` is called before `onSubscribe(Subscription s)` it seems that the NPE would still occurs, should that be a concern?

Reply:
I don't think it would since inputStreamState is UNINITIALIZED and it will execute the first if block